### ISSUE TITLE
Direction: Mamba/SSM as Tier 2 replacement for transformer

### ISSUE THREAD

**Issue by travis-mcdevitt** [travis-mcdevitt/parameter-golf#5]
## Thesis

The transformer's KV cache is the single biggest constraint on eval-time prediction quality at 16MB. Replacing the transformer with a **state-space model (Mamba)** eliminates the KV cache entirely, enabling full-context evaluation over the entire validation set with O(1) memory per token.

This is the highest-EV architectural change that's executable within realistic engineering constraints.

## The eval-time bottleneck nobody talks about

Current SOTA uses sliding-window eval at stride=64 with seq_len=2048. This means:
- Each token is predicted using at most 2048 tokens of context
- 97% of context computation is **redundant** — overlapping windows recompute the same prefixes
- Eval takes ~77-105 seconds out of a 600 second budget
- Extending context to 4096 roughly doubles eval time, eating the remaining budget

A Mamba model with O(1) recurrent state processes the entire validation set in a **single left-to-right pass**. No sliding windows, no redundant computation, no KV cache. Every token sees the full preceding context.

## Why this might work at tiny scale

The conventional wisdom is "transformers beat SSMs at language modeling." But that result comes from large-scale comparisons (1B+ parameters) on diverse benchmarks. At 20M parameters on web text compression:

1. **Short-range patterns dominate.** At tiny scale, most of the model's capacity goes to local predictions anyway. Mamba's recurrent inductive bias is well-suited for sequential structure.
2. **The inference advantage compounds.** Full-context eval means the model sees patterns that sliding-window transformers literally cannot — cross-document regularities, long-range formatting patterns, repeated boilerplate.
3. **No embedding table advantage for transformers.** Both architectures pay the same vocab×dim embedding cost. The savings are in the sequence-processing layers.
4. **Mamba layers are parameter-efficient.** A Mamba block has ~3×D×D_inner parameters (vs ~4×D² + 2×D×D_mlp for a transformer block). At the same parameter budget, you can fit more Mamba layers.

### Parameter budget comparison (16MB, int6 quantized)

| Architecture | Layers | Params | Context at eval |
|-------------|--------|--------|----------------|
| Transformer (SOTA) | 11 | ~24M | 2048 (sliding window) |
| Mamba (same budget) | ~14-16 | ~24M | Unlimited (recurrent) |

### Phase 1: Direct replacement smoke test

Replace the transformer blocks in the tiered predictor (#2) with Mamba-2 blocks. Keep everything else identical — same tokenizer, same tiers 0+1, same optimizer split, same quantization.

```
Tier 0: learned per-token bias (unchanged)
Tier 1: BigramHash (unchanged)  
Tier 2: Mamba-2 stack (replaces transformer)
```

The Mamba-2 selective scan has production CUDA kernels via `mamba_ssm`. Training speed on H100 is competitive with FlashAttention transformers.

### Phase 2: Full-context eval

Modify the eval loop:
- Instead of sliding-window batches, process the validation set as a single long sequence
- Carry the Mamba hidden state forward across the entire eval
- No padding, no overlap, no wasted compute

### Phase 3: Architecture tuning

Once the baseline works:
- Sweep depth vs width tradeoff (more narrow layers vs fewer wide layers)
- Test hybrid Mamba-attention (a few attention layers for long-range, mostly Mamba for efficiency)
- Explore state size vs quality tradeoff (larger SSM state = better memory but more parameters)

## Training considerations

**Mamba training is parallel.** Despite being a recurrent model at inference, Mamba-2 trains using a parallel scan that processes full sequences in one pass. Training throughput on H100 is within ~20% of an equivalently-sized transformer with FlashAttention.

**Optimizer compatibility.** Muon (Newton-Schulz orthogonalization) works on any 2D weight matrix. Mamba's input/output projections are standard linear layers. The SSM-specific parameters (A, B, C, D matrices, dt) are small and can use Adam. The existing `SplitOptimizers` pattern carries over directly.

**Not PyTorch-dependent.** The `mamba_ssm` package provides CUDA kernels. For MLX smoke testing, a pure-Python selective scan implementation exists (slow but functional). For H100 training, the CUDA path is production-ready.

## What could go wrong

1. **Quality gap at tiny scale.** Mamba might simply predict worse than a transformer at 20M params. The recurrent bottleneck (fixed-size state) limits information flow in ways that attention doesn't. This is the central risk.
2. **Training instability.** SSMs can be finicky with learning rates and initialization. The Newton-Schulz orthogonalization in Muon hasn't been tested with SSM parameters.
3. **Quantization behavior.** SSM state matrices (A, B, C) have different numerical properties than attention weights. Int6 GPTQ might not work as well. Needs testing.
4. **MLX support.** No native Mamba kernels in MLX. Smoke testing on Mac requires a pure-Python fallback, which will be slow.

## Feasibility assessment

| Aspect | Rating | Notes |
|--------|--------|-------|
| Ceiling | High | Full-context eval is structurally better than sliding window |
| Training feasibility | Good | `mamba_ssm` CUDA kernels exist, training speed is competitive |
| Engineering effort | Medium | Drop-in replacement for transformer blocks, eval loop change |
| Eval-time feasibility | Excellent | O(1) state, single-pass eval, no KV cache |
| Risk | Medium | Quality gap at tiny scale is unknown |

## Relationship to other directions

- Direct replacement for Tier 2 in the tiered predictor (#2)
- Orthogonal to MDL smoothness (#3) — Mamba weights can also benefit from smoothness reg
- Complementary to context mixing (forthcoming) — Mamba could be the neural component in a learned context mixer
- Strictly better than eval-time context extension (#1) — full context by default, no RoPE extrapolation needed

## Next steps

1. Build a minimal Mamba-2 block in the MLX tiered predictor for smoke testing
2. Benchmark training throughput on H100 vs transformer baseline
3. Compare val_bpb at matched parameter count and training time
4. If competitive: implement full-context eval and measure the BPB gain from unlimited context
